business/layer.go

```diff
@@ -80,7 +82,13 @@ func Get(token string) (*Layer, error) {
 		prometheusClient = prom
 	}
 
-	return NewWithBackends(k8s, prometheusClient), nil
+	// Create Jaeger client
```

Comment:
I think this is not optimal because we're creating a jaeger client every time someone makes an access to the business layer (even when it's not going to use jaeger at all). Probably we should do some lazy loading instead.

As you wish, I'm fine if you prefer not doing it in this PR, but then we should create a ticket to not forget about that, wdyt?

Reply:
Ok  @jotak , I added this like a client thinking about to improve integrations with more endpoints or providers, let me fix it

Reply:
If user with admin rights requests client, this is stored with his token in the instance, if another user request with another rights what happen?  How kiali deal with this in prometheus? Am I missing code?

Reply:
With prometheus it's a bit different because Kiali has its own method to grant permission for metrics per user, so we always use Kiali's own service account token and never the user one. You can see that mentioned here: https://github.com/kiali/kiali/blob/master/prometheus/client.go#L51-L53 . Also in the kiali_cr documentation:

```yaml
#   use_kiali_token: When true and if auth.type is "bearer", Kiali Service Account token will be used for the API calls to Prometheus
```

whereas for Grafana or Jaeger it says:

```yaml
#   use_kiali_token: When true and if auth.type is "bearer", the same OAuth token used for authentication in Kiali will be used for the API calls to Grafana,
```

So that makes the things a bit easier to code for Prometheus.

Reply:
@aljesusg I think we could have lazy loading that deals with the token by capturing it via a loader function. I can give some help if you want

Reply:
That could be something around these lines:

- In Jaeger service:

```go
type JaegerLoader = func() (jaeger.ClientInterface, error)

type JaegerService struct {
	loader        JaegerLoader
	loaderErr     error
	jaeger        jaeger.ClientInterface
	businessLayer *Layer
}
```

Then we can create a function `client()` that performs the lazy loading. When the service wants to get the client, we can do:

```go
func (in *JaegerService) client() (jaeger.ClientInterface, error) {
	if in.jaeger != nil {
		return in.jaeger, nil
	} else if in.loaderErr != nil {
		return nil, in.loaderErr
	}
	in.jaeger, in.loaderErr = in.loader()
	return in.jaeger, in.loaderErr
}

// In GetJaegerServices for example:
	client, err := in.client()
	if err != nil {
		return nil, 0, err
	}
	return client.GetJaegerServices()
```

- And from layer.go to create the service with the loader, using the captured token:

```go
// ...
	// Create Jaeger client loader
	loader := func() (jaeger.ClientInterface, error) {
		return jaeger.NewClient(token)
	}
	return NewWithBackends(k8s, prometheusClient, loader), nil
```

And `NewWithBackends` creates the service with just the loader:

```go
        JaegerService{loader: jaegerLoader, businessLayer: temporaryLayer}
```